Implement initial framework for wamp transports

davidchappelle opened the pull request:
This commit removes the templatization of the wamp session. This
will allow us to:

- Treat transports and sessions as separate entities which allows for:
    - Reuse of transports across multiple sessions.
    - Pausing and resuming of sessions.
- Implement different types of transports (i.e. websocket).
- Implement flow control for transports.
- Support native communication with the application router.

jpetso commented on autobahn/wamp_session.hpp:
```diff
@@ -211,15 +222,16 @@ class wamp_session : public std::enable_shared_from_this<wamp_session<IStream, O
 
     /*!
-     * Register an procedure as a procedure that can be called remotely.
+     * Register a that can be called remotely.
```
Something's missing here :)

jpetso commented on autobahn/wamp_session.hpp:
```diff
@@ -228,15 +240,22 @@ class wamp_session : public std::enable_shared_from_this<wamp_session<IStream, O
 
     /*!
-     * Function called by the session when authenting. It always has to be re-implemented (if authentication is part of the system).
+     * Function called by the session when authenting. It always has to be
```
Likely not your typo, but if you're already reflowing these comments then you might as well change `authenting` to `authenticating`.

jpetso commented on autobahn/wamp_session.hpp:
```diff
@@ -228,15 +240,22 @@ class wamp_session : public std::enable_shared_from_this<wamp_session<IStream, O
      *
-     * \param challenge The challenge from the router. Containing enough information, for the system to prove membership.
-     * \return A future deliver a proper authentication response to the challenge given.
+     * \param challenge The challenge from the router. Contains enough information,
```
In the same vein, there's a superfluous comma here.

jpetso commented on autobahn/wamp_rawsocket_transport.hpp:
```diff
@@ -0,0 +1,237 @@
+///////////////////////////////////////////////////////////////////////////////
+//
+// Copyright (c) Tavendo GmbH
```
Given the extent of your changes, I'm pretty sure we should add Topology copyright to all of those files that you're adding or rewriting (in addition to Tavendo copyright), e.g.:
`Copyright (c) 2015, 2016 Topology LP`

@oberstet, I suggest you add years for Tavendo as well, otherwise it gets hard to tell when the copyright expires. (This is making the entirely reasonable & unquestionable assumption that in 70 or 95 years or however long copyright will be then, autobahn-cpp will be in wide use as a foundational piece of all software.)

davidchappelle replied:
I am going to add a Topology copyright to every file as we have made significant modifications to all files. I will do it in a separate commit.

I think the preferred approach is to not include a copyright date as it generally just causes confusion to those not familiar with copyright law. The commit meta-data contains dates anyways. plus it will inevitably get out of sync as the code is updated.

oberstet commented:
Regarding copyright: I'd be fine with replacing the line

```
// Copyright (c) Tavendo GmbH
```

with

```
// Copyright (c) Tavendo GmbH, Topology LP and contributors.
```

in all existing files, and add the line to all new files.

> We haven't added explicit other contributor names to any of the other Autobahn libraries, but I think the amount of code contributed by Topology, and the specific situation here definitely justifies that. (I am cautious, because I want to avoid being asked to do similar for X and Y and Z on all Autobahn's - I wouldn't!). 

I agree with @davidchappelle regarding dates (no dates pls), exactly because of the arguments Dave raises. In any case, everything can be traced back to the first commit anyway - and those commits have timestamps.

oberstet commented:
@jpetso 

> This is making the entirely reasonable & unquestionable assumption that in 70 or 95 years or however long copyright will be then, autobahn-cpp will be in wide use as a foundational piece of all software.

=)

I totally agree.

oberstet commented:
Btw: why are messages shuffled around `.. got_message(wamp_message&& message)`  with double `&&`?

I can see these types coming from code

```cpp
wamp_message message(std::move(fields), std::move(*(result.zone())));
got_message(std::move(message));
```

I still don't get it. And reading [this](http://stackoverflow.com/a/14486705/884770) makes my brain hurt;)

---

A minor thing .. but. I recently took time making sure all the examples run, work with the Crossbar.io config supplied in the examples dir out of the box, and that the READMEs are in sync.

I've skimmed over the patch (that also updates the examples) .. and I think there are some minor issues. I will test the branch using Crossbar.io from that directory. Yes, I understand you (obviously) care more about Bonefish (and that is perfectly fine!), but people come and file issues and ask questions on our mailing lists if AutobahnCpp does not work with Crossbar.io and/or the examples here.

---

I'll also wait for the copyright changes before merging - if Topology is fine with the proposal above.

davidchappelle commented:
The double && allows for ownership of objects to be passed efficiently without having to deal with smart pointers, heap allocations, etc. It took a bit once I moved to c++11 to really see the advantages to r-value references. In the code you posted above, the message is constructed from a set of fields and a zone. The fields and zone only exist for the purposes of constructing the message. To avoid the overhead of copying, heap allocations, etc. we use r-value references. The calls to std::move(...) are effectively zero-cost cast operations. It allows for the r-value versions of the copy and assignment constructors to be invoked where appropriate. This also allows the transports themselves to deal with messages efficiently. For example, when back pressure occurs or when native communication is being used (i.e. autobahn-cpp calling directly into the application router and vice versa).

Hopefully things still just work with crossbar. I think the only sticking point would be the uds example but  I added a new test parameter for specifying the uds path. Maybe I should default to the crossbar default?

I am in agreement with how to do the copyright. I can put together a commit for that once @jpetso confirms.

oberstet commented:
@davidchappelle thanks for explaining the double & thing. I have filed https://github.com/crossbario/autobahn-cpp/issues/94 as I think this is worth discussing in docs.

Rgd examples:

* the examples include https://github.com/crossbario/autobahn-cpp/blob/master/examples/web/index.html, and there URIs there should match the one's used in C++.
* As of the UDS path: whatever the default from command line arg parsing should match https://github.com/crossbario/autobahn-cpp/blob/master/examples/.crossbar/config.json#L79 - we can make that more generic though, eg `router.sock` or whatever.

oberstet commented:
@davidchappelle pls let me know if you still want this merged, given the different views on "copyright notice" in the headers ..

davidchappelle commented:
Yes I am ok with still merging. I am just touching up the example items
that you mentioned above.

On Wed, Feb 17, 2016 at 4:35 PM, Tobias Oberstein <notifications@github.com>
wrote:

> @davidchappelle <https://github.com/davidchappelle> pls let me know if
> you still want this merged, given the different views on "copyright notice"
> in the headers ..
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/crossbario/autobahn-cpp/pull/93#issuecomment-185416831>
> .
>

oberstet commented:
Ok, cool!! Thanks! Gimme a ping when ready and I'll merge.

oberstet commented:
Thanks a lot, Dave!

davidchappelle commented:
Not a problem. Thanks for all of the collaboration. I think that
autobahn-cpp is now at a stage where many of the newer serialization and
transport features can be easily added.

On Wed, Feb 17, 2016 at 4:58 PM, Tobias Oberstein <notifications@github.com>
wrote:

> Thanks a lot, Dave!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/crossbario/autobahn-cpp/pull/93#issuecomment-185425547>
> .
>

DZabavchik commented:
Great job @davidchappelle!  Been tracking your `wamp_transports` branch for a while.
